kubeadm: fix RequiredIPVSKernelModulesAvailableCheck

bart0sh opened the pull request:
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

kubeadm IPVS check confuses users as it doesn't
take into account installed modules. It only checks loaded
and builtin modules. If it can't find required module among
loaded or/and builtin modules it fails.
However, it shouldn't be considered an error as installed
modules can still be loaded.

Modified RequiredIPVSKernelModulesAvailableCheck code to
also check installed modules using 'modinfo <module>'

**Which issue(s) this PR fixes**:

Fixes: kubernetes/kubeadm#975

**Does this PR introduce a user-facing change?**:

```release-note
kubeadm: Fixed IPVS check to consider installed kernel modules
```

bart0sh commented:
/cc @neolit123

neolit123 commented:
/area kubeadm
@kubernetes/sig-network-pr-reviews 
@kubernetes/sig-cluster-lifecycle-pr-reviews

neolit123 commented on pkg/util/ipvs/kernelcheck_linux.go:
```diff
@@ -75,18 +75,21 @@ func (r RequiredIPVSKernelModulesAvailableCheck) Check() (warnings, errors []err
-				builtInModules.Insert(string(builtInMode))
+				out, err = r.Executor.Command("modinfo", module).CombinedOutput()
+				if err != nil {
+					warnings = append(warnings, fmt.Errorf("error getting %s module info: %v: %s", module, err, strings.TrimSpace(string(out))))
```
`error getting %s module info: %v: %s`
->
`error getting module info for %q; error: %v; output: %s`

and a related question, do we benefit from CombinedOutput() here?
can you please show example output?

if not we can omit `; output: %s`

bart0sh replied:
I think we do:
```
# modinfo bla
modinfo: ERROR: Module bla not found.
$ modinfo bla
-bash: modinfo: command not found
```

neolit123 commented on pkg/util/ipvs/kernelcheck_linux.go:
```diff
@@ -75,18 +75,21 @@ func (r RequiredIPVSKernelModulesAvailableCheck) Check() (warnings, errors []err
+			match, _ := regexp.Match(module+".ko", out)
 			if !match {
-				builtInModules.Insert(string(builtInMode))
+				out, err = r.Executor.Command("modinfo", module).CombinedOutput()
```
any noticeable performance drop given the number of modules we need to check?

bart0sh replied:
worst case scenario(no loaded and no builtin modules found):
```
# time sh -c "for m in ip_vs ip_vs_rr ip_vs_wrr ip_vs_sh nf_conntrack_ipv4; do modinfo \$m ; done"
...
real	0m0.004s
user	0m0.004s
sys	0m0.000s
```

neolit123 commented on pkg/util/ipvs/kernelcheck_linux.go:
```diff
@@ -75,18 +75,21 @@ func (r RequiredIPVSKernelModulesAvailableCheck) Check() (warnings, errors []err
-					"you can solve this problem with following methods:\n 1. Run 'modprobe -- ' to load missing kernel modules;\n"+
-					"2. Provide the missing builtin kernel ipvs support\n", modules, builtInModules))
+				"the IPVS proxier will not be used, because the following required kernel modules are not loaded or found: %v\n"+
+					"You can solve this problem by providing builtin or loadable kernel ipvs support\n", missingModules))
```
`kernel ipvs support` -> `kernel IPVS support`

neolit123 reviewed (commented):
thanks @bart0sh 
added some minor comments / questions.

bart0sh commented:
@neolit123 thank you for the review!

kvaps commented:
We probably should fix also kube-proxy, there is similar check, because it's disables IPVS if required kernel modules not found

m1093782566 commented:
Thanks @neolit123 

A quick question: are you using IPVS proxier?

cc @islinwb @stewart-yu @Lion-Wei 

for review.

m1093782566 commented:
Thanks @neolit123 

A quick question: are you using IPVS proxier?

cc @islinwb @stewart-yu @Lion-Wei 

for review.

stewart-yu commented:
In the function, first part is `checking installed modules`.
And, why delete the builtin mode check?

islinwb commented:
> We probably should fix also kube-proxy, there is similar check, because it's disables IPVS if required kernel modules not found

I don't think the problem exists in kube-proxy. Maybe you can submit an issue.

kvaps commented:
> And, why delete the builtin mode check?

Why is it needed, if all modules are existing and working well, just not in listed in `modules.builtin`?

Example, my case:
https://github.com/kubernetes/kubeadm/issues/975#issuecomment-430990149

bart0sh commented:
@stewart-yu 
> In the function, first part is checking installed modules.

It's checking loaded modules as far as I understand the code.

> And, why delete the builtin mode check?
It's not deleted. I've just added check for installed modules.

I've tested this solution. When ip_vs* modules are not builtin and not loaded before the kubeadm run they're loaded automatically if they're installed. The same reported in the issue: 
https://github.com/kubernetes/kubeadm/issues/975#issuecomment-405738687

bart0sh commented on pkg/util/ipvs/kernelcheck_linux.go:
```diff
@@ -75,18 +75,21 @@ func (r RequiredIPVSKernelModulesAvailableCheck) Check() (warnings, errors []err
+			match, _ := regexp.Match(module+".ko", out)
 			if !match {
-				builtInModules.Insert(string(builtInMode))
+				out, err = r.Executor.Command("modinfo", module).CombinedOutput()
```
it wouldn't as far as I understood. If they're not loaded and not installed they're still be built into the kernel. We should do all 3 checks in my opinion.

bart0sh replied:
That makes sense. Will do. Thank you for the idea!

stewart-yu replied:
any update?

bart0sh replied:
@stewart-yu I'm travelling. will update this PR today/tomorrow.

bart0sh commented:
/test pull-kubernetes-integration

bart0sh commented:
@stewart-yu done. please review.

stewart-yu commented on pkg/util/ipvs/kernelcheck_linux.go:
```diff
@@ -47,7 +47,7 @@ func (r RequiredIPVSKernelModulesAvailableCheck) Check() (warnings, errors []err
 	if err != nil {
-		errors = append(errors, fmt.Errorf("error getting installed ipvs required kernel modules: %v(%s)", err, out))
+		errors = append(errors, fmt.Errorf("error getting installed ipvs required kernel modules: %v(%s)", err, string(out)))
 		return nil, errors
```
should we change this?

bart0sh replied:
no, we shouldn't. reverted back. Thanks.

stewart-yu commented on pkg/util/ipvs/kernelcheck_linux_test.go:
```diff
@@ -30,65 +31,74 @@ func TestRequiredIPVSKernelModulesAvailableCheck(t *testing.T) {
 			kernelVersion:  "3.13.0-24-generic",
 			builtinKernel:  "",
+			modinfoError:   fmt.Errorf("modinfo error"),
 			expectErrors:   false,
```
why `modinfoError` not nil?

bart0sh replied:
I think I already answered this. Answering again:
It's not nil as this test case expects loadable modules to be installed.

stewart-yu commented on pkg/util/ipvs/kernelcheck_linux_test.go:
```diff
@@ -30,65 +31,74 @@ func TestRequiredIPVSKernelModulesAvailableCheck(t *testing.T) {
-				"kernel/net/netfilter/ipvs/ip_vs_sh.ko\n" +
-				"kernel/net/ipv4/netfilter/nf_conntrack_ipv4.ko",
+				"kernel/net/netfilter/ipvs/ip_vs_wrr.ko\n",
+			modinfoError:   nil,
```
why `modinfoError` nil?

stewart-yu commented on pkg/util/ipvs/kernelcheck_linux.go:
```diff
@@ -58,37 +58,54 @@ func (r RequiredIPVSKernelModulesAvailableCheck) Check() (warnings, errors []err
-			return nil, errors
+			warnings = append(warnings, fmt.Errorf("error getting module info for %q; error: %v; output: %s", module, err, strings.TrimSpace(string(out))))
+			missingModules = append(missingModules, module)
 		}
```
Actually, i'm do not clear why should add this check. can you explain more details?

bart0sh replied:
sure. we run modinfo <module> to check if loadable modules are installed. If modinfo returns non-zero return code there could be different reasons for that, e.g. there is no modinfo tool in the PATH, there is no info about the module etc. It makes sense to show those reasons to users if modules are not found in my opinion. If they're found among builtin modules those warnings are ignored.

stewart-yu reviewed (commented):
looks better now, some nil comments

bart0sh commented:
@stewart-yu can you review this please?

bart0sh committed:
kubeadm: fix RequiredIPVSKernelModulesAvailableCheck

kubeadm IPVS check confuses users as it doesn't take into account
installed modules. It only checks loaded and builtin modules.
If it can't find required module among loaded and builtin modules
it fails. However, it shouldn't be considered an error as installed
modules can still be loaded.

Modified RequiredIPVSKernelModulesAvailableCheck code to
also check installed modules using 'modinfo <module>'

Fixes: kubernetes/kubeadm#975

bart0sh commented:
Rebased on top of the latest master.

@neolit123, @stewart-yu please, review.

neolit123 commented on pkg/util/ipvs/kernelcheck_linux.go:
```diff
@@ -63,29 +64,46 @@ func (r RequiredIPVSKernelModulesAvailableCheck) Check() (warnings, errors []err
 		if err != nil {
-			errors = append(errors, fmt.Errorf("error getting required builtin kernel modules: %v(%s)", err, out))
-			return nil, errors
+			warnings = append(warnings, fmt.Errorf("error getting module info for %q; error: %v; output: %s", module, err, strings.TrimSpace(string(out))))
```
we decided to use pkg-error everywhere in the meantime:
https://godoc.org/github.com/pkg/errors

-->
```
errors.Wrapf(err, "error getting module info for %q; output: %s; error:", module, strings.TrimSpace(string(out)))
```

neolit123 commented on pkg/util/ipvs/kernelcheck_linux.go:
```diff
@@ -63,29 +64,46 @@ func (r RequiredIPVSKernelModulesAvailableCheck) Check() (warnings, errors []err
+	builtinModsFilePath := fmt.Sprintf("/lib/modules/%s/modules.builtin", kernelVersion)
+	out, err = r.Executor.Command("cut", "-f1", "-d", " ", builtinModsFilePath).CombinedOutput()
+	if err != nil {
+		errors = append(errors, fmt.Errorf("error getting required builtin kernel modules: %v(%s)", err, out))
```
errors.Wrapf()
out should go before err.

neolit123 commented on pkg/util/ipvs/kernelcheck_linux.go:
```diff
@@ -63,29 +64,46 @@ func (r RequiredIPVSKernelModulesAvailableCheck) Check() (warnings, errors []err
 
-	return warnings, errors
+	if len(missingBuiltins) != 0 {
+		warnings = append(warnings, fmt.Errorf(
```
errors.Errorf()

stewart-yu commented:
@kvaps would you like to verify the step again? https://github.com/kubernetes/kubeadm/issues/975#issuecomment-430990149

bart0sh commented:
@stewart-yu 
> Actually, I confused about what this PR fix is?

Would [this comment](
https://github.com/kubernetes/kubeadm/issues/975#issuecomment-432153976) make you less confused?

bart0sh commented:
/cc kad

kvaps commented:
> The describe in [kubernetes/kubeadm#975 (comment)](https://github.com/kubernetes/kubeadm/issues/975#issuecomment-430990149) also confused me.

Hi, there was my mistake I wasn't load modules by proper way: 
https://github.com/kubernetes/kubeadm/issues/975#issuecomment-432158733
probably, this `lsmod` output made after initialization. Sorry for confusing, I was confused myself :)

Currently kubeadm will display warning if listed modules not loaded in the system.
And it will show no warning if all modules loaded proper way.

>Actually, I confused about what this PR fix is?

This PR disables warning if listed modules not loaded.
But enables check are they existing.

Another words, if user have all necessary modules, he will not see this warning anymore, even if they are not loaded into the system.
Kube-proxy will load those modules automatically, on demand.

bart0sh commented:
/test pull-kubernetes-verify

stewart-yu commented:
> Kube-proxy will load those modules automatically, on demand.

No need to load those modules automatically. https://github.com/kubernetes/kubernetes/pull/59566#issuecomment-390212614

> Another words, if user have all necessary modules, he will not see this warning anymore, even if they are not loaded into the system.

If user have all necessary modules, he will not see this warning anymore, that's true. And we have this function now.

bart0sh commented:
@stewart-yu
> If user have all necessary modules, he will not see this warning anymore, that's true.

That's absolutely not true. I have all required modules installed:

```
$ find /lib/modules/$(uname -r) -name ip_vs.ko -o -name ip_vs_rr.ko -o -name ip_vs_wrr.ko -o -name ip_vs_sh.ko -o -name nf_conntrack_ipv4.ko
/lib/modules/4.4.140-62-default/kernel/net/ipv4/netfilter/nf_conntrack_ipv4.ko
/lib/modules/4.4.140-62-default/kernel/net/netfilter/ipvs/ip_vs_wrr.ko
/lib/modules/4.4.140-62-default/kernel/net/netfilter/ipvs/ip_vs.ko
/lib/modules/4.4.140-62-default/kernel/net/netfilter/ipvs/ip_vs_rr.ko
/lib/modules/4.4.140-62-default/kernel/net/netfilter/ipvs/ip_vs_sh.ko
```

However, the check still makes incorrect and confusing  warning message:
```
I1113 13:18:01.781297   14832 kernelcheck_linux.go:45] validating the kernel module IPVS required exists in machine or not
	[WARNING RequiredIPVSKernelModulesAvailable]: the IPVS proxier will not be used, because the following required kernel modules are not loaded: [ip_vs_rr ip_vs_wrr ip_vs_sh ip_vs] or no builtin kernel ipvs support: map[ip_vs:{} ip_vs_rr:{} ip_vs_wrr:{} ip_vs_sh:{} nf_conntrack_ipv4:{}]
you can solve this problem with following methods:
 1. Run 'modprobe -- ' to load missing kernel modules;
2. Provide the missing builtin kernel ipvs support
```

This is plain wrong: ```the IPVS proxier will not be used```

As IPVS proxier is used just fine after ```kubeadm init``` is finished:
```
$ sudo /sbin/ipvsadm -ln
IP Virtual Server version 1.2.1 (size=4096)
Prot LocalAddress:Port Scheduler Flags
  -> RemoteAddress:Port           Forward Weight ActiveConn InActConn
TCP  10.96.0.1:443 lc
  -> 10.237.72.179:6443           Masq    1      0          0         
TCP  10.96.0.10:53 lc
UDP  10.96.0.10:53 lc
```

Therefore both recommendations:
```
1. Run 'modprobe -- ' to load missing kernel modules;
2. Provide the missing builtin kernel ipvs support
```
don't make sense and meaningless.

@neolit123, @kad, @timothysc please, help me to get this PR merged. It's stuck under review for 24 days already. It's a valid fix for kubernetes/kubeadm#975

kvaps commented:
@stewart-yu 

> >    If user have all necessary modules, he will not see this warning anymore, that's true.
>
>That's absolutely not true. I have all required modules installed:

Agree, warning still there if modules not loaded but existing on the system.


> >   Kube-proxy will load those modules automatically, on demand.
>
>No need to load those modules automatically. #59566 (comment)

kube-proxy already doing that, we don't need to load them on preflight check.

kvaps commented:
@stewart-yu 
The problem is that modules can be on the machine but not listed on modules.builtin file. This PR removes warning in this case.

bart0sh commented:
@stewart-yu 
> If not mind, could you like ping me in Slack? we can discuss more detail

sure. Which channel are you at? What's your nick? I couldn't find you there.

bart0sh commented:
> Secondary, loaded those file in system, like command

This is a main point of misunderstanding. You don't need to do any 'modprobe -- <name>' to be able to work with IPVS. Current implementation of RequiredIPVSKernelModulesAvailable check insist on that and this is confusing. That was the reason for this issue: kubernetes/kubeadm#975 and this PR.

bart0sh commented:
@Lion-Wei It still looks confusing to me. Why to throw a warning that can be ignored in most of the cases and everything would work just fine?

neolit123 commented:
@bart0sh please close this PR as it seems the PR replaces it.
thanks.

kad commented:
> @bart0sh @kvaps
> I know kube-proxy can execute `modprobe ...`, but when kube-proxy running inside container, then we need privilege to run this command, that is insecurity, and we are trying to revoke the privilege of kube-proxy container.
kube-proxy anyway executed in privileged mode as it need access to control host kernel IPVS and iptables rules.

kad commented on pkg/util/ipvs/kernelcheck_linux.go:
```diff
@@ -63,29 +65,46 @@ func (r RequiredIPVSKernelModulesAvailableCheck) Check() (warnings, errors []err
+
+	// Check if builtin modules exist
+	builtinModsFilePath := fmt.Sprintf("/lib/modules/%s/modules.builtin", kernelVersion)
+	out, err = r.Executor.Command("cut", "-f1", "-d", " ", builtinModsFilePath).CombinedOutput()
```
Can we get rid of this by replacing it with simple read+parse function?

bart0sh replied:
I'd be happy to do that if we reach overall agreement to accept this patch with @Lion-Wei and @stewart-yu

bart0sh replied:
@neolit123 

> please close this PR as it seems the PR replaces it.

Let's give @Lion-Wei and @stewart-yu another chance to accept it. I still think it's a correct fix and after @kad's comment I don't see any reason of not accepting it.

cmluciano replied:
+1 would be great if we can use some of the existing util functions for checking these i.e. https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/iptables/proxier.go#L129-L138

bart0sh commented:
@neolit123

> please close this PR as it seems the PR replaces it.

Let's give @Lion-Wei and @stewart-yu another chance to accept it. I still think it's a correct fix and after @kad's comment I don't see any reason of not accepting it.

bart0sh commented:
/assign @stewart-yu

bart0sh commented:
@stewart-yu kindly ping for review&acceptance. Please consider last comment from @kad.

cmluciano commented:
@bart0sh Can you please address the remaining comments and rebase this PR so we can continue reviewing?

bart0sh commented:
@cmluciano

>  Can you please address the remaining comments and rebase this PR so we can continue reviewing?

Can you please review [this pr](https://github.com/kubernetes/kubernetes/pull/75036) instead?
I'm going to deprecate this if that one is approved.

cmluciano commented:
@bart0sh  Reviewed, please close this one out

bart0sh commented:
@cmluciano 

>  Reviewed, please close this one out

Will do after that one will be approved.

bart0sh commented:
closed as https://github.com/kubernetes/kubernetes/pull/75036 has been merged
